envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -140,41 +140,56 @@ class RBACFilterFactory(
         return methodPermissions
     }
 
-    private fun resolveClients(
+    private fun resolveClientsWithSelectors(
         incomingEndpoint: IncomingEndpoint,
         roles: List<Role>
-    ): Collection<ClientMatching> {
+    ): Collection<ClientWithSelector> {
         val clients = incomingEndpoint.clients.flatMap { clientOrRole ->
-            roles.find { it.name == clientOrRole.name }?.clients?.map { it } ?: setOf(clientOrRole)
-        }
+            roles.find { it.name == clientOrRole.name }?.clients ?: setOf(clientOrRole)
+        }.toSortedSet()
         // sorted order ensures that we do not duplicate rules
-        return clients.toSortedSet(Comparator.comparing<ClientMatching, String> { it.name })
+        return clients
     }
 
     private fun mapMethodToHeaderMatcher(method: String): Permission {
         val methodMatch = HeaderMatcher.newBuilder().setName(":method").setExactMatch(method).build()
         return Permission.newBuilder().setHeader(methodMatch).build()
     }
 
-    private fun mapClientMatchingToPrincipals(
-        clientMatching: ClientMatching,
+    private fun mapClientWithSelectorToPrincipals(
+        clientWithSelector: ClientWithSelector,
         snapshot: GlobalSnapshot
     ): List<Principal> {
-        val staticRangesForClient = staticIpRange(clientMatching)
+        val selectorMatching = getSelectorMatching(clientWithSelector, incomingPermissionsProperties)
+        val staticRangesForClient = staticIpRange(clientWithSelector, selectorMatching)
 
-        return if (clientMatching.name in incomingServicesSourceAuthentication) {
-            ipFromDiscoveryPrincipals(clientMatching, snapshot)
+        return if (clientWithSelector.name in incomingServicesSourceAuthentication) {
+            ipFromDiscoveryPrincipals(clientWithSelector.name, clientWithSelector.selector, selectorMatching, snapshot)
```

Comment:
Maybe it would be nicer to pass just `clientWithSelector` as an argument, instead of its fields.